tree.go

```diff
@@ -22,33 +21,34 @@ type Tree interface {
 	Root() ([]byte, error)
 }
 
-var _ Tree = &DefaultTree{}
+// trees is a global map used for keeping track of registered tree constructors for JSON serialization
+// The keys of this map should be kebab cased. E.g. "default-tree"
+var trees = make(map[string]TreeConstructorFn)
+
+func RegisterTree(treeName string, treeConstructor TreeConstructorFn) error {
+	if trees[treeName] != nil {
+		return fmt.Errorf("%s already registered", treeName)
+	}
+	trees[treeName] = treeConstructor
 
-type DefaultTree struct {
-	*merkletree.Tree
-	leaves [][]byte
-	root   []byte
+	return nil
 }
 
-func NewDefaultTree(_ Axis, _ uint) Tree {
-	return &DefaultTree{
-		Tree:   merkletree.New(sha256.New()),
-		leaves: make([][]byte, 0, 128),
+func GetTreeConstructorFn(treeName string) (TreeConstructorFn, error) {
```

Comment:
A little doc won't hurt here. I feel like the name is too verbose, but that's because of the type. I would call it TreeFn, but we don't want to break things here